### PR TITLE
Flow trigger CLI and defaults 

### LINKED ISSUE #4686: trigger: generalisation of triggering approaches
> **Related Issues:**
>
> If agreed this issue should supersede:
> * https://github.com/cylc/cylc-flow/issues/4657
> * https://github.com/cylc/cylc-flow/issues/4653

After a long chat with @dpmatthews (who proposed yet another triggering approach 😁) I think we can generalise the trigger problem into two dimensions:

* Continue (yes/no).
  * After I trigger the task will the flow continue from that point immediately.
  * Or does it only continue if/when a flow front catches up with it.
  * I.E. Should the triggered tasks spawn children on completion or after "merge".
* Overrun (yes/no).
  * Should the "merge" [1] condition be based on the pool or the DB?
  * I.E. Should triggered tasks overrun previous runs of tasks?
  * I.E. Should the following flow overrun the triggered tasks?

> **Note:** From the internal implementation these two dimensions may appear flip-sides of the same coin since they both boil down to the `flow_nums`, however, considering them from a user standpoint I think it's fair to prise them apart.

> **Note:** Purposefully using new terminology to avoid conflation with existing terms, we may want to workshop "continue" and "overrun" a touch.

> **[1]:** The quoted "merge" above relates to the interaction between two tasks with different `flow_nums` in general and *not* to the more specific concept of "flow merging" in the pool exclusively.

Combing these we get four spaces:

|            | Continue                 | Don't Continue                                          |
|------------|-----------------------|---------------------------------------------------|
| Overrun    | (1) Reflow (as currently implemented)               | (3) No Flow (current default trigger behaviour)                  |
| No Overrun | (2) Continue (@dpmatthews new proposed implementation) | (4) No Flow (@oliver-sanders [proposed implementation](https://github.com/cylc/cylc-flow/issues/4657)) |

* The bad news is it looks like we have use cases for all four.
* Dave & I think the no-overrun cases are more important than the overrun ones.
* The good news is that they can coexist and the mechanism for supporting all four is currently implemented, it's mostly an interface problem.

### 1) Reflow (implemented)

Equivalent to `cylc trigger --flow=<new-flow-number>`.

*Continue: Yes*
*Overrun: Yes*

* Tasks are triggered with a new flow number.
* The reflow can overrun previous flows.
* The reflow will merge if it collides with another flow in the pool (and only in the pool i.e. overrun).

The use case is for re-running over tasks which have been previously run e.g. change configuration and re-run a sub-graph.

### 2) Continue (proposed)

Equivalent to `cylc trigger --flow=<all-flow-numbers>,<new-flow-number>`.

*Continue: Yes*
*Overrun: No*

* A new trigger approach proposed by @dpmatthews.
* Tasks are triggered with *all* existing flow numbers *plus* a new flow number (which we added purely so the new flow can still be targeted by CLI tools).
* Because this flow contains all existing flow numbers it will not be overrun by *any* of the flows which exist at the time of the trigger.
* This is intended for the sort of use cases we would expect `--flow=1` to be used for, but has been generalised to be reflow compatible.

This approach feels quite "natural". The use cases are setting off another bit of the same flow where you don't want tasks to be overrun.

### 3) No Flow (implemented)

Equivalent to `cylc trigger --flow -1`.

> I am using a negative flow number rather than `None` to distinguish the two no-flow approaches.
> Internally we can still maintain the same no-flow logic as present but would need to change the marker.

*Continue: No*
*Overrun: Yes*

Useful for running one-off tasks that you do not want to impact the workflow in any way (i.e. `cylc submit` type uses).

### 4) No Flow (proposed)

Equivalent to `cylc trigger --flow -2`.

> I am using a negative flow number rather than None to distinguish the two no-flow approaches.
> Internally we can still maintain the same no-flow logic as present.

*Continue: No*
*Overrun: No*

Use case is for manually intervening in graph execution by ignoring dependencies or runahead limit and skipping ahead to a task which you want to be considered a part of the approaching flow front.

## Interface

The internals to handle the four cases are already in-place, flow_nums, DB lookups etc, so it mostly boils down to an interface / documentation issue.

I think all four methods *could* be exposed via a single `--flow` argument, however, it is sensible to provide defaults for the different behaviours. I think it would be good to document the `--flow` equivalents as they may help users to understand their function.

> Note that `--reflow` currently determines the new flow number server rather than client side which is sensible.

### 1) Enable behaviours explicitly

If we are happy with the continue/overrun model (after workshopping the terms) we could expose it directly something like:

```bash
# 1) reflow
cylc trigger --continue --overrun

# 2) continue
cylc trigger --continue

# 3) no-flow (implemented)
cylc trigger --overrun

# 4) no-flow (proposed)
cylc trigger
```

This is quite nice as you have to explicitly opt in to each behaviour separately reducing the scope for unintended results and accidents.

### 2) Single `--flow` argument

if we don't like the continue/overrun model we could move the presets into the flow argument something like:

```bash
# 1) reflow
cylc trigger --flow=new

# 2) continue
cylc trigger --flow=any

# 3) no-flow (implemented)
cylc trigger --flow=none

# 4) no-flow (proposed)
cylc trigger --flow=next
```

It's less behaviour driven so we would need to explain each option separately.

### 3) Separate flag for each approach

An alternative to (2) would be to  could come up with three/four different flags:

```bash
# 1) reflow
cylc trigger --reflow

# 2) continue
cylc trigger --flow

# 3) no-flow (implemented)
cylc trigger --rerun

# 4) no-flow (proposed)
cylc trigger  # --run
```

### Default

I think no-continue & no-overrun is the safest, sanest default because:

* The minimum set of behaviours is the simplest.
* The "Continue" cases have a dramatic impact on the workflow execution and are hard to revoke.
* The "Re-run" cases are quite advanced and require additional knowledge to operate.

But I'm biased. I think the default is less important than the clear separation of behaviours.

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #4686.  Built on top of - and now replaces - #4738 

Implement final agreed CLI and defaults for flow triggering.

~~TBD~~
- [x] currently spawns on all outputs of the triggered task on continuing the flow; use actual completed outputs
- [x] make it work even if the same task is triggered multiple times
- [x] tests for `--flow=INT`

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.

- [x] Appropriate tests are included (unit and/or functional).

- [x] Appropriate change log entry included.

- [x] (master branch) I have opened a documentation PR at  https://github.com/cylc/cylc-doc/pull/427

### CODE REVIEW

**Comment by hjoliver** [comment]
Fast tests will fail till #4746 goes in.

**Comment by hjoliver** [comment]
This fixes and extends the DB `task_outputs` table. On master, the table holds only custom outputs, for the most recent submit of each task. Here, it holds all outputs, with flow numbers. E.g.:

```console
$ sqlite3 -header ~/cylc-run/ft/runN/log/db "select * from task_outputs;"
cycle|name|flow_nums|outputs
1|foo|[1]|["submitted", "started", "succeeded"]
1|bar|[1]|["submitted", "started", "the quick brown fox", "succeeded"]
1|baz|[1]|["submitted", "started", "failed"]
```
For repeat-running a task *in the same flow* via `cylc trigger --wait` we care about outputs from the most recent run in the right flow, not necessarily the most recent submit.  So at this stage I don't think it's necessary to add submit number to the table as well.

**Comment by hjoliver** [comment]
(One last TODO for tomorrow: correctly handle `n=0` flow merge with a still-running `trigger --wait` task)

**Comment by hjoliver** [comment]
(Rebased and deconflicted)

**Comment by hjoliver** [comment]
>Here's a PR that adds a default value to the flow arg to allow tui to work with the mutation:

Thanks, merged!

File: cylc/flow/task_pool.py [hunk 1]
```diff
@@ -1453,16 +1482,22 @@ def force_trigger_tasks(
...
+                if flow[0] == FLOW_ALL:
+                    flow_nums = self._get_active_flow_nums()
+                elif flow[0] == FLOW_NEW:
                     flow_nums = {self.flow_mgr.get_new_flow(flow_descr)}
-                else:
+                elif flow[0] == FLOW_NONE:
                     flow_nums = set()
-                itask = self.spawn_task(name, point, flow_nums)
+                else:
+                    flow_nums = {int(n) for n in flow}
```
Comment (oliver-sanders) [inline cylc/flow/task_pool.py:1494]: I think this gets pre-validated(?) but should still probably still stick this in a try/except for safety as a ValueError here would tear down the workflow.

File: cylc/flow/task_pool.py [hunk 2]
```diff
@@ -1453,16 +1482,22 @@ def force_trigger_tasks(
...
+                itask = self.spawn_task(name, point, flow_nums, force=True)
                 if itask is None:
                     continue
+                # TODO PUT THESE IN TASK INIT
```
Comment (oliver-sanders) [inline cylc/flow/task_pool.py:1498]: TODO

**Review by oliver-sanders — COMMENTED** [review]
Most the way through.

File: cylc/flow/task_pool.py [hunk 3]
```diff
@@ -422,7 +423,7 @@ def load_db_task_pool_for_restart(self, row_idx, row):
                     TASK_STATUS_FAILED,
                     TASK_STATUS_SUCCEEDED
             ):
-                for message in json.loads(outputs_str).values():
+                for message in json.loads(outputs_str):
                     itask.state.outputs.set_completion(message, True)
                     self.data_store_mgr.delta_task_output(itask, message)
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:428]: No. On master, we store both the label and the message for custom outputs.  On this branch, just the message.

**Comment by hjoliver** [comment]
Feedback addressed.

**Comment by dwsutherland** [comment]
> > Here's a PR that adds a default value to the flow arg to allow tui to work with the mutation:
> 
> Thanks, merged!

Doesn't look like it's merged..

**Comment by hjoliver** [comment]
> > > Here's a PR that adds a default value to the flow arg to allow tui to work with the mutation:
> > 
> > 
> > Thanks, merged!
> 
> Doesn't look like it's merged..

Oops, good spot. It is now. I must have been waiting for tests or something.

**Comment by hjoliver** [comment]
Imported a flake8-bugbear error from @oliver-sanders side-PR:
> ./cylc/flow/network/schema.py:1410:5: B021 f-string used as docstring.This will be interpreted by python as a joined string rather than a docstring.

**Comment by hjoliver** [comment]
(OK you bet me to it :-)

**Comment by hjoliver** [comment]
Rebased.

**Comment by hjoliver** [comment]
(Damn it, tests all passed before I merged that side PR and rebased!)

**Comment by hjoliver** [comment]
> Imported a flake8-bugbear error from @oliver-sanders side-PR:
> 
> > ./cylc/flow/network/schema.py:1410:5: B021 f-string used as docstring.This will be interpreted by python as a joined string rather than a docstring.

FYI @oliver-sanders (and others) - doc-strings cannot be f-strings:

https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals

**Comment by hjoliver** [comment]
All fixed :tada:

**Review by dwsutherland — APPROVED** [review]
Looks good to me:
- Worked in my environment.
- Tests passed.
- Code is fine.

👍

**Comment by oliver-sanders** [comment]
> FYI @oliver-sanders (and others) - doc-strings cannot be f-strings:

TIL. You can modify, format or even replace a docstring after it has been declared, however, you cannot do so at the time of declaration.

```python
FOO = 42


class Modified:
    '{FOO}'
    

Modified.__doc__ = Modified.__doc__.format(**globals())


class FString:
    f'''{FOO}'''
    
    
class Format:
    '''{FOO}'''.format(**globals())


# this works
print(Modified.__doc__)

# these don't
print(FString.__doc__)
print(Format.__doc__)
```

```console
$ python mess.py 
42
None
None
```

**Review by oliver-sanders — CHANGES_REQUESTED** [review]
Completed functional review, all good except one bug which can cause multiple new flows to be created for a single trigger `--flow=new` event.

Put up a fix: https://github.com/hjoliver/cylc-flow/pull/22

Otherwise good!

File: cylc/flow/network/schema.py [hunk 4]
```diff
@@ -1404,6 +1406,11 @@ def description(self):
         return StopMode(self.value).describe()
 
 
+class Flow(String):
+    """An integer or one of {FLOW_ALL}, {FLOW_NEW} or {FLOW_NONE}."""
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/network/schema.py:1410]: We can't expand them, or the docstring ceases to be a docstring (see my comment on that).

I think it's fine to leave like that. Docstrings describe code, so it's not unreasonable for them to refer to variable (or constant) names. I should perhaps have removed the braces, but I added to comment to note that it shouldn't be turned back into an f-string.

**Review by oliver-sanders — APPROVED** [review]
:tada:

Thanks @hjoliver this has been a fun one!

**Comment by hjoliver** [comment]
> tada
> 
> Thanks @hjoliver this has been a fun one!

Heheh, yes it has :rofl: 

:tada:

**Pull request merged**